tensorboard/components/tf_backend/urlPathHelpers.ts

```diff
@@ -12,17 +12,42 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-export function queryEncoder(params?: any): string {
-  // It's important that the keys be sorted, so we always grab the right file
-  // if we are talking to the backend generated by serialze_tensorboard.py
-  if (params == null) {
-    return '';
+/**
+ * A query parameter value can either be a string or a list of strings.
+ * A string `"foo"` is encoded as `key=foo`; a list `["foo", "bar"]` is
+ * encoded as `key=foo&key=bar`.
+ */
+export type QueryValue = string | string[];
+
+/**
+ * Add query parameters to a URL. Values will be URL-encoded. The URL
+ * may or may not already have query parameters. For convenience,
+ * parameters whose value is `undefined` will be dropped.
+ *
+ * For example, the following expressions are equivalent:
+ *
+ *     addParams("http://foo", {a: "1", b: ["2", "3+4"], c: "5"})
+ *     addParams("http://foo?a=1", {b: ["2", "3+4"], c: "5", d: undefined})
+ *     "http://foo?a=1&b=2&b=3%2B4&c=5"
+ */
+export function addParams(
+    baseURL: string,
+    params: {[param: string]: QueryValue}): string {
+  const keys = Object.keys(params).sort().filter(k => params[k] !== undefined);
+  if (!keys.length) {
+    return baseURL;  // no need to change '/foo' to '/foo?'
   }
-  const components = _.keys(params)
-                       .sort()
-                       .filter((k) => params[k] !== undefined)
-                       .map((k) => k + '=' + encodeURIComponent(params[k]));
-  const result = components.length ? '?' + components.join('&') : '';
-  // Replace parens for consistency with urllib.urlencode
-  return result.replace(/\(/g, '%28').replace(/\)/g, '%29');
+  const delimiter = baseURL.indexOf('?') !== -1 ? '&' : '?';
+  const parts = [].concat(...keys.map(key => {
+    const rawValue = params[key];
+    const values = Array.isArray(rawValue) ? rawValue : [rawValue];
+    return values.map(value => `${key}=${_encodeURIComponent(value)}`);
+  }));
+  const query = parts.join('&');
+  return baseURL + delimiter + query;
+}
+
+function _encodeURIComponent(x: string): string {
+  // Replace parentheses for consistency with Python's urllib.urlencode.
```

Comment:
Nice. That's subtle.

Reply:
Yep. (Also included in previous `queryEncoder`.) Less actually useful now that we don't care about demo mode, but I figured that we might as well keep it around.